src/util/small_shared_ptr.h

```diff
@@ -0,0 +1,228 @@
+/*******************************************************************\
+
+Module: Small intrusive shared pointers
+
+Author: Reuben Thomas, reuben.thomas@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_SMALL_SHARED_PTR_H
+#define CPROVER_UTIL_SMALL_SHARED_PTR_H
+
+#include <iosfwd>  // ostream
+#include <utility> // swap
+
+// TODO We should liberally scatter `constexpr`s and `noexcept`s on platforms
+// that support them.
+
+// This class is really similar to boost's intrusive_pointer, but can be a bit
+// simpler seeing as we're only using it one place...
+
+template <typename T>
+class small_shared_ptrt final
+{
+public:
+  small_shared_ptrt() = default;
+
+  explicit small_shared_ptrt(T *t) : t_(t)
+  {
+    if(t_)
+    {
+      pointee_increment_use_count(*t_);
+    }
+  }
+
+  small_shared_ptrt(const small_shared_ptrt &rhs) : t_(rhs.t_)
+  {
+    if(t_)
+    {
+      pointee_increment_use_count(*t_);
+    }
+  }
+
+  small_shared_ptrt &operator=(const small_shared_ptrt &rhs)
+  {
+    auto copy(rhs);
+    swap(copy);
+    return *this;
+  }
+
+  small_shared_ptrt(small_shared_ptrt &&rhs)
+  {
+    swap(rhs);
+  }
+
+  small_shared_ptrt &operator=(small_shared_ptrt &&rhs)
+  {
+    swap(rhs);
+    return *this;
+  }
+
+  ~small_shared_ptrt()
+  {
+    if(!t_)
+    {
+      return;
+    }
+    if(pointee_use_count(*t_) == 1)
+    {
+      delete t_;
+      return;
+    }
+    pointee_decrement_use_count(*t_);
```

Comment:
How about:
````c++
if(t_)
{
  if(pointee_use_count(*t_) == 1)
    delete t_;
  else
    pointee_decrement_use_count(*t_);
}
````

Reply:
I don't really like nesting blocks too much if I can avoid it. I'm leaving this as-is in the absence of a coding guideline on the topic.

Reply:
Sure, our personal tastes probably just differ on this one :-)